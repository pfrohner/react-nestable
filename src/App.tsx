import React, { FC, useState } from 'react';
import Nestable from 'react-nestable';
import { uid } from 'react-uid'
import { Formik, ErrorMessage  } from 'formik';
import * as yup from 'yup';
import filterDeep from 'deepdash/filterDeep';

import { TextField, Button } from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/Delete';

import logo from './logo.svg';
import './App.css';

type Item = {
  id: string;
  text: string;
  children?: Item[]
}

type FormData = {
  text: string;
}

const initialValues: FormData = {
  text: 'John'
}

const validationSchema = yup.object({
  text: yup
    .string()
    .required()
    .min(4)
    .max(50)
});

const App: FC = () => {
  const [items, setItems] = useState<Item[]>([]);

  const onSubmit = (data: FormData, actions: { setSubmitting: (state: boolean) => void }) => {
    actions.setSubmitting(true)
    const newItem: Item = {
      id: uid(data),
      text: data.text
    }
    setItems([...items, newItem]);
    data.text = '';
    actions.setSubmitting(false);
  }
  const onDelete = (id: string): void => {
    const isSure: boolean = window.confirm('Are you sure?');
    if (isSure) {
      const filteredItems: Item[] = filterDeep(items, (value: { id: string; }) => value.id !== id,
      { childrenPath: 'children', pathFormat: 'array' }
      )
      setItems(filteredItems ?? []);
    }
  }
  const onNestableChange = (newState: Item[]) => setItems(newState);
  const renderItem = ({ item }: { item: { id: string, text: string, children: [] }}): any => {
    return <>
      {item.text}
      {item.children.length === 0 && <DeleteIcon
        fontSize="small"
        className="DeleteIcon"
        onClick={() => onDelete(item.id)} />
      }
    </>
  }

  return (
    <div className="App">
      <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        <p>React Nestable demo</p>
      </header>
      <main className="Content">
        <div className="Form">
          <Formik
            initialValues={initialValues}
            onSubmit={onSubmit}
            validateOnChange={false}
            validationSchema={validationSchema}
          >
            {({ values, isSubmitting, handleChange, handleBlur, handleSubmit }) => (
              <form onSubmit={handleSubmit}>
                <TextField
                  name="text"
                  value={values.text}
                  onChange={handleChange}
                  placeholder="Add new item"
                  onBlur={handleBlur}  />
                <Button
                  disabled={isSubmitting}
                  type="submit"
                  size="small"
                  variant="contained"
                  color="primary"
                  className="SubmitButton"
                >
                  Submit
                </Button>
                <div>
                  <ErrorMessage name="text" />
                </div>
              </form>
            )}
          </Formik>
        </div>
        <Nestable
          onChange={onNestableChange}
          items={items}
          renderItem={renderItem}
          maxDepth={3}
        />
      </main>
    </div>
  );
}

export default App;
