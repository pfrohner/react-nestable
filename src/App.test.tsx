import React from 'react';
import { render } from '@testing-library/react';
import App from './App';

test('rendering one item', () => {
  const { getByText } = render(<App />);
  const testElement = getByText(/john/i);
  
  expect(testElement).toBeInTheDocument();
});
