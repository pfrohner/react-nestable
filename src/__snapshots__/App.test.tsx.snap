// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
      >
        <header
          class="App-header"
        >
          <img
            alt="logo"
            class="App-logo"
            src="logo.svg"
          />
          <p>
            React Nestable demo
          </p>
        </header>
        <main
          class="Content"
        >
          <div
            class="Form"
          >
            <form>
              <div
                class="MuiFormControl-root MuiTextField-root"
              >
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    name="text"
                    placeholder="Add new item"
                    type="text"
                    value="John"
                  />
                </div>
              </div>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained SubmitButton MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
                tabindex="0"
                type="submit"
              >
                <span
                  class="MuiButton-label"
                >
                  Submit
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <div />
            </form>
          </div>
          <div
            class="nestable nestable-g3546wwjggh"
          >
            <ol
              class="nestable-list nestable-group"
            />
          </div>
        </main>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
    >
      <header
        class="App-header"
      >
        <img
          alt="logo"
          class="App-logo"
          src="logo.svg"
        />
        <p>
          React Nestable demo
        </p>
      </header>
      <main
        class="Content"
      >
        <div
          class="Form"
        >
          <form>
            <div
              class="MuiFormControl-root MuiTextField-root"
            >
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiInput-input"
                  name="text"
                  placeholder="Add new item"
                  type="text"
                  value="John"
                />
              </div>
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained SubmitButton MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
              tabindex="0"
              type="submit"
            >
              <span
                class="MuiButton-label"
              >
                Submit
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
            <div />
          </form>
        </div>
        <div
          class="nestable nestable-g3546wwjggh"
        >
          <ol
            class="nestable-list nestable-group"
          />
        </div>
      </main>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
